File: uiweb/src/control/MiniCCommandRegistry.ts
import { MiniCControlCommand } from "./MiniCControlCommand";
import type { JavaMirrorFile } from "../translation/javaMirror";
import { requireValue } from "../translation/uiTypes";

export const miniCCommandRegistryMirror = {
  "javaPath": "src/main/java/minic/uilocal/control/MiniCCommandRegistry.java",
  "webPath": "uiweb/src/control/MiniCCommandRegistry.ts",
  "packageName": "minic.uilocal.control",
  "exportName": "MiniCCommandRegistry",
  "kind": "class",
  "imports": [
    "java.util.LinkedHashMap",
    "java.util.Map",
    "java.util.Objects",
    "java.util.Optional"
  ],
  "fields": [
    {
      "name": "commands",
      "signature": "private final Map<String,MiniCControlCommand>commands="
    }
  ],
  "methods": [
    {
      "name": "command",
      "signature": "command(String id)"
    },
    {
      "name": "enabled",
      "signature": "enabled(String id)"
    },
    {
      "name": "execute",
      "signature": "execute(String id)"
    },
    {
      "name": "register",
      "signature": "register(MiniCControlCommand command)"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCCommandRegistry {
  static readonly mirror = miniCCommandRegistryMirror;

  readonly mirror = miniCCommandRegistryMirror;

  private readonly commands = new Map<string, MiniCControlCommand>();

  register(command: MiniCControlCommand): void {
    const safeCommand = requireValue(command, "command");
    this.commands.set(safeCommand.id(), safeCommand);
  }

  command(id: string): MiniCControlCommand | undefined {
    return this.commands.get(id);
  }

  enabled(id: string): boolean {
    return this.command(id)?.enabled()() ?? false;
  }

  execute(id: string): boolean {
    const command = this.commands.get(id);
    if (!command || !command.enabled()()) {
      return false;
    }
    command.action()();
    return true;
  }

  commandIds(): readonly string[] {
    return [...this.commands.keys()];
  }

  summary(): string {
    return `MiniCCommandRegistry: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCCommandRegistry;

File: src/main/java/minic/uilocal/control/MiniCCommandRegistry.java
package minic.uilocal.control;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class MiniCCommandRegistry {
    private final Map<String, MiniCControlCommand> commands = new LinkedHashMap<>();

    public void register(MiniCControlCommand command) {
        Objects.requireNonNull(command, "command");
        commands.put(command.id(), command);
    }

    public Optional<MiniCControlCommand> command(String id) {
        return Optional.ofNullable(commands.get(id));
    }

    public boolean enabled(String id) {
        return command(id)
                .map(MiniCControlCommand::enabled)
                .map(supplier -> supplier.getAsBoolean())
                .orElse(false);
    }

    public boolean execute(String id) {
        MiniCControlCommand command = commands.get(id);
        if (command == null || !command.enabled().getAsBoolean()) {
            return false;
        }
        command.action().run();
        return true;
    }
}
